int-set/src/bitset.rs

```diff
@@ -43,21 +43,34 @@ impl<T: Into<u32> + Copy> BitSet<T> {
         for major in major_start..=major_end {
             let page_start = start.max(self.major_start(major));
             let page_end = end.min(self.major_start(major + 1) - 1);
-            self.page_for_major_mut(major)
+            self.ensure_page_for_major_mut(major)
                 .insert_range(page_start, page_end);
         }
         self.mark_dirty();
     }
 
+    /// An alternate version of extend() which is optimized for inserting an unsorted
+    /// iterator of values.
+    pub fn extend_unsorted<U: IntoIterator<Item = T>>(&mut self, iter: U) {
```

Comment:
this is slightly confusing, since we wouldn't want `extend` to ever assume that input was sorted.


My gut then would be to invert this, and have `extend` be naive, and have `extend_sorted` be the special additional function.


At a higher level, I would really prefer if we could have a single implementation that just switched behaviour after seeing the first out-of-order item?

Reply:
The current extend() implementation works fine for unsorted or partial sorted inputs, but will be slightly slower (~7%) then this version. When input is sorted or partially sorted however there will be a massive speedup (84% in the benchmark), so I think the right tradeoff is to have that as the default and leave extend_unsorted() as a niche optimization. The way extend() is currently written there's no need to bail out to an alternate implementation when encountering unsorted data. This allows is to provide speedups for partially sorted inputs as well.